File: app/src/main/java/net/osdn/gokigen/pkremote/MainActivity.java
package net.osdn.gokigen.pkremote;

import android.Manifest;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.graphics.Typeface;
import android.os.Bundle;

import androidx.annotation.NonNull;

import com.google.android.material.bottomnavigation.BottomNavigationView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.preference.PreferenceManager;

import android.os.Vibrator;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.view.WindowManager;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ImageButton;
import android.widget.Spinner;
import android.widget.TextView;

import net.osdn.gokigen.pkremote.camera.CameraInterfaceProvider;
import net.osdn.gokigen.pkremote.camera.interfaces.IInterfaceProvider;
import net.osdn.gokigen.pkremote.camera.interfaces.control.ICameraConnection;
import net.osdn.gokigen.pkremote.preference.IPreferencePropertyAccessor;
import net.osdn.gokigen.pkremote.scene.CameraSceneUpdater;

/**
 *
 *
 */
public class MainActivity extends AppCompatActivity implements View.OnClickListener, IInformationReceiver, ICardSlotSelector, AdapterView.OnItemSelectedListener
{
    private final String TAG = toString();
    private IInterfaceProvider interfaceProvider = null;
    private CameraSceneUpdater scenceUpdater = null;

    private ImageButton mImageConnectButton = null;
    private ImageButton mReloadButton = null;
    private Spinner mCardSlotSelection = null;
    private ICardSlotSelectionReceiver slotSelectionReceiver = null;
    //private TextView mTextMessage = null;

    private BottomNavigationView.OnNavigationItemSelectedListener mOnNavigationItemSelectedListener
            = new BottomNavigationView.OnNavigationItemSelectedListener() {

        @Override
        public boolean onNavigationItemSelected(@NonNull MenuItem item) {
            switch (item.getItemId()) {
                case R.id.navigation_photo_library:
                    scenceUpdater.changeScenceToImageList();
                    return (true);
                case R.id.navigation_calendar:
                    scenceUpdater.changeSceneToCalendar();
                    return (true);
                case R.id.navigation_auto_transfer:
                    scenceUpdater.changeSceneToAutoTransfer();
                    return (true);
                case R.id.navigation_settings:
                    scenceUpdater.changeSceneToConfiguration();
                    return (true);
            }
            return (false);
        }
    };

    /**
     *
     *
     */
    @Override
    public void onBackPressed()
    {
        //Log.v(TAG, "onBackPressed()");
        super.onBackPressed();
        runOnUiThread(new Runnable() {
            @Override
            public void run()
            {
                if (scenceUpdater != null)
                {
                    scenceUpdater.updateBottomNavigationMenu();
                }
            }
        });
    }

    /**
     *
     *
     */
    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        try
        {
            ActionBar bar = getSupportActionBar();
            if (bar != null)
            {
                // タイトルバーは表示しない
                bar.hide();
            }
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);

        //mTextMessage = findViewById(R.id.message);
        mImageConnectButton = findViewById(R.id.button_wifi_connect);
        mReloadButton = findViewById(R.id.button_reload);
        mCardSlotSelection = findViewById(R.id.card_slot_selection);

        BottomNavigationView navigation = findViewById(R.id.navigation);
        navigation.setOnNavigationItemSelectedListener(mOnNavigationItemSelectedListener);

        // パーミッション群のオプトイン
        final int REQUEST_NEED_PERMISSIONS = 1010;

        if ((ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) ||
                (ContextCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) ||
                (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_MEDIA_LOCATION) != PackageManager.PERMISSION_GRANTED) ||
                (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_NETWORK_STATE) != PackageManager.PERMISSION_GRANTED) ||
                (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_WIFI_STATE) != PackageManager.PERMISSION_GRANTED) ||
                (ContextCompat.checkSelfPermission(this, Manifest.permission.VIBRATE) != PackageManager.PERMISSION_GRANTED) ||
                (ContextCompat.checkSelfPermission(this, Manifest.permission.INTERNET) != PackageManager.PERMISSION_GRANTED)) {
            ActivityCompat.requestPermissions(this,
                    new String[]{
                            Manifest.permission.WRITE_EXTERNAL_STORAGE,
                            Manifest.permission.READ_EXTERNAL_STORAGE,
                            Manifest.permission.ACCESS_MEDIA_LOCATION,
                            Manifest.permission.ACCESS_NETWORK_STATE,
                            Manifest.permission.ACCESS_WIFI_STATE,
                            Manifest.permission.VIBRATE,
                            Manifest.permission.INTERNET,
                    },
                    REQUEST_NEED_PERMISSIONS);
        }
        initializeClass();
        prepareClass();
        onReadyClass();
    }

    /**
     * パーミッション設定が終わった後...
     */
    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String permissions[], @NonNull int[] grantResults)
    {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        prepareClass();
        onReadyClass();
    }

    /**
     *
     */
    @Override
    protected void onPause()
    {
        super.onPause();
        try
        {
            ICameraConnection connection = interfaceProvider.getCameraConnection();
            if (connection != null)
            {
                connection.stopWatchWifiStatus(this);
            }
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    /**
     * クラスの初期化 (instantiate)
     */
    private void initializeClass()
    {
        try
        {
            scenceUpdater = CameraSceneUpdater.newInstance(this);
            interfaceProvider = CameraInterfaceProvider.newInstance(this, scenceUpdater, this, this);
            scenceUpdater.changeFirstFragment(interfaceProvider);
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    /**
     * 本クラスの準備
     */
    private void prepareClass()
    {
        try
        {
            mImageConnectButton.setOnClickListener(this);
            mReloadButton.setOnClickListener(this);
            setupCardSlotSelection(interfaceProvider.getCammeraConnectionMethod() == ICameraConnection.CameraConnectionMethod.PANASONIC);
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    private void setupCardSlotSelection(boolean isEnabled)
    {
        try
        {
            if (mCardSlotSelection == null)
            {
                mCardSlotSelection = findViewById(R.id.card_slot_selection);
            }
            if (isEnabled)
            {
                // 接続モードが Panasonic の時だけ、SD Card 選択を出せるようにする
                mCardSlotSelection.setVisibility(View.VISIBLE);
                ArrayAdapter<CharSequence> adapter = ArrayAdapter.createFromResource(this, R.array.sd_card_slot, android.R.layout.simple_spinner_item);
                adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
                mCardSlotSelection.setAdapter(adapter);
                mCardSlotSelection.setOnItemSelectedListener(this);
            }
            else
            {
                mCardSlotSelection.setVisibility(View.GONE);
            }
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }



    /**
     * 初期化終了時の処理 (カメラへの自動接続)
     */
    private void onReadyClass()
    {
        try
        {
            // カメラに自動接続するかどうか確認
            SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);
            boolean isAutoConnectCamera = preferences.getBoolean(IPreferencePropertyAccessor.AUTO_CONNECT_TO_CAMERA, true);
            Log.v(TAG, "isAutoConnectCamera() : " + isAutoConnectCamera);

            // カメラに接続する
            if (isAutoConnectCamera)
            {
                // 自動接続の指示があったとき
                scenceUpdater.changeCameraConnection();
            }
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    @Override
    public void onClick(View v)
    {
        try
        {
            int id = v.getId();
            switch (id)
            {
                case R.id.button_wifi_connect:
                    // カメラとの接続を行う
                    scenceUpdater.changeCameraConnection();
                    vibrate();
                    break;

                case R.id.button_reload:
                    // 画像一覧情報をリロードする
                    scenceUpdater.reloadRemoteImageContents();
                    vibrate();
                    break;

                default:
                    break;
            }
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    /**
     *
     *
     */
    private void vibrate()
    {
        try {
            Vibrator vibrator = (Vibrator) getSystemService(VIBRATOR_SERVICE);
            if (vibrator != null)
            {
                vibrator.vibrate(50);
            }
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }


    @Override
    public void updateMessage(final String message, final boolean isBold, final boolean isColor,  final int color)
    {
        try {
            final TextView messageArea = findViewById(R.id.message);
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    try {
                        if ((messageArea != null) && (message != null))
                        {
                            messageArea.setText(message);
                            if (isBold)
                            {
                                messageArea.setTypeface(Typeface.DEFAULT_BOLD);
                            }
                            if (isColor)
                            {
                                messageArea.setTextColor(color);
                            }
                            else
                            {
                                messageArea.setTextColor(Color.DKGRAY);
                            }
                            messageArea.invalidate();
                        }
                    }
                    catch (Exception e)
                    {
                        e.printStackTrace();
                    }
                }
            });
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    @Override
    public void onItemSelected(AdapterView<?> parent, View view, int position, long id)
    {
        try
        {
            String item = (String) parent.getItemAtPosition(position);
            Log.v(TAG, " onItemSelected : " +  item);
            if (slotSelectionReceiver != null)
            {
                slotSelectionReceiver.slotSelected(item);
            }
            vibrate();

            // scenceUpdater.reloadRemoteImageContents();
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    @Override
    public void onNothingSelected(AdapterView<?> parent)
    {
        // 何もしない
    }

    @Override
    public void setupSlotSelector(final boolean isEnable, @Nullable ICardSlotSelectionReceiver slotSelectionReceiver)
    {
        try
        {
            Log.v(TAG, "  ------- setupSlotSelector " + isEnable);
            this.slotSelectionReceiver = slotSelectionReceiver;
            runOnUiThread(new Runnable() {
                @Override
                public void run()
                {
                    setupCardSlotSelection(isEnable);
                }
            });
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/net/osdn/gokigen/pkremote/MainActivity.kt
package net.osdn.gokigen.pkremote

import android.Manifest
import android.content.pm.PackageManager
import android.graphics.Color
import android.graphics.Typeface
import android.os.Build
import android.os.Bundle
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import android.util.Log
import android.view.View
import android.view.WindowManager
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.updatePadding
import androidx.preference.PreferenceManager
import net.osdn.gokigen.pkremote.camera.CameraInterfaceProvider
import net.osdn.gokigen.pkremote.camera.interfaces.IInterfaceProvider
import net.osdn.gokigen.pkremote.camera.interfaces.control.ICameraConnection
import net.osdn.gokigen.pkremote.camera.interfaces.status.ICameraStatusReceiver
import net.osdn.gokigen.pkremote.preference.IPreferencePropertyAccessor
import net.osdn.gokigen.pkremote.scene.CameraSceneUpdater
import androidx.activity.OnBackPressedCallback
import androidx.activity.result.contract.ActivityResultContracts
import net.osdn.gokigen.pkremote.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity(),
    IInformationReceiver,
    ICardSlotSelector
{
    private lateinit var binding: ActivityMainBinding

    private var interfaceProvider: IInterfaceProvider? = null
    private var sceneUpdater: CameraSceneUpdater? = null
    private var slotSelectionReceiver: ICardSlotSelectionReceiver? = null

    // --- ActivityResult APIを使用
    private val requestPermissionLauncher = registerForActivityResult(
        ActivityResultContracts.RequestMultiplePermissions()
    ) { permissions ->
        val allGranted = permissions.entries.all { it.value }
        if (allGranted) {
            try {
                prepareClass()
                onReadyClass()
            }
            catch (_: Exception)
            {
                Log.e(TAG, "class initialize fail...")
            }
        } else {
            Log.v(TAG, "Permissions not granted")
            Toast.makeText(this, R.string.permission_not_granted, Toast.LENGTH_SHORT).show()
            finish()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?)
    {
        super.onCreate(savedInstanceState)
        Log.v(TAG, " ----- onCreate() -----")

        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        supportActionBar?.hide()
        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)

        setupWindowInset(binding.container)
        setupNavigation()
        setupOnBackPressed()

        initializeClass()

        checkAndRequestPermissions()
    }

    private fun setupOnBackPressed() {
        // onBackPressed 対策
        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                isEnabled = false
                onBackPressedDispatcher.onBackPressed()
                sceneUpdater?.updateBottomNavigationMenu()
            }
        })
    }

    private fun setupNavigation() {
        binding.navigation.setOnItemSelectedListener { item ->
            when (item.itemId) {
                R.id.navigation_photo_library -> {
                    sceneUpdater?.changeScenceToImageList()
                    true
                }
                R.id.navigation_calendar -> {
                    sceneUpdater?.changeSceneToCalendar()
                    true
                }
                R.id.navigation_auto_transfer -> {
                    sceneUpdater?.changeSceneToAutoTransfer()
                    true
                }
                R.id.navigation_settings -> {
                    sceneUpdater?.changeSceneToConfiguration()
                    true
                }
                else -> false
            }
        }
    }

    // OSバージョンに応じて必要な権限のみを取得してチェックする
    private fun checkAndRequestPermissions() {
        val requiredPermissions = getRequiredPermissions()
        val missingPermissions = requiredPermissions.filter {
            ContextCompat.checkSelfPermission(this, it) != PackageManager.PERMISSION_GRANTED
        }.toTypedArray()

        if (missingPermissions.isEmpty()) {
            Log.v(TAG, "allPermissionsGranted() : true")
            prepareClass()
            onReadyClass()
        } else {
            Log.v(TAG, "====== REQUEST PERMISSIONS ======")
            requestPermissionLauncher.launch(missingPermissions)
        }
    }

    private fun getRequiredPermissions(): List<String> {
        val list = mutableListOf(
            Manifest.permission.INTERNET,
            Manifest.permission.VIBRATE,
            Manifest.permission.ACCESS_NETWORK_STATE,
            Manifest.permission.ACCESS_WIFI_STATE
        )

        // Android 12 (API 31) 以下のみ必要なストレージ権限
        if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.S_V2) {
            list.add(Manifest.permission.READ_EXTERNAL_STORAGE)
            list.add(Manifest.permission.WRITE_EXTERNAL_STORAGE)
        }

        // Android 10 (API 29) 以降
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            list.add(Manifest.permission.ACCESS_MEDIA_LOCATION)
        }

        // Android 13 (API 33) 以降
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            list.add(Manifest.permission.NEARBY_WIFI_DEVICES)
        }

        // Android 17 (API 37) 以降
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.CINNAMON_BUN) {
            list.add(Manifest.permission.ACCESS_LOCAL_NETWORK)
        }

        return list
    }

    private fun setupWindowInset(view: View)
    {
        ViewCompat.setOnApplyWindowInsetsListener(view) { v, insets ->
            val bars = insets.getInsets(
                WindowInsetsCompat.Type.systemBars() or WindowInsetsCompat.Type.displayCutout()
            )
            v.updatePadding(
                left = bars.left,
                top = bars.top,
                right = bars.right,
                bottom = bars.bottom
            )
            WindowInsetsCompat.CONSUMED
        }
    }

    override fun onPause()
    {
        super.onPause()
        try { interfaceProvider?.cameraConnection?.stopWatchWifiStatus(this) } catch (_: Exception) { }
    }

    private fun initializeClass()
    {
        sceneUpdater = CameraSceneUpdater.newInstance(this)
        sceneUpdater?.let { updater ->
            val scene = updater as ICameraStatusReceiver
            val provider = CameraInterfaceProvider.newInstance(this, scene, this, this)
            interfaceProvider = provider
            updater.changeFirstFragment(provider)
        }
    }

    private fun prepareClass()
    {
        Log.v(TAG, "prepareClass()")

        binding.buttonWifiConnect.setOnClickListener {
            sceneUpdater?.changeCameraConnection()
            vibrate()
        }

        binding.buttonReload.setOnClickListener {
            sceneUpdater?.reloadRemoteImageContents()
            vibrate()
        }

        val isPanasonic = interfaceProvider?.cammeraConnectionMethod == ICameraConnection.CameraConnectionMethod.PANASONIC
        setupCardSlotSelection(isPanasonic)
    }

    private fun setupCardSlotSelection(isEnabled: Boolean) {
        if (isEnabled)
        {
            binding.cardSlotSelection.visibility = View.VISIBLE
            val adapter = ArrayAdapter.createFromResource(
                this,
                R.array.sd_card_slot,
                android.R.layout.simple_spinner_item
            ).apply {
                setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
            }
            binding.cardSlotSelection.adapter = adapter
            binding.cardSlotSelection.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>, view: View?, position: Int, id: Long) {
                    val item = parent.getItemAtPosition(position) as String
                    slotSelectionReceiver?.slotSelected(item)
                }

                override fun onNothingSelected(parent: AdapterView<*>?) {}
            }
        }
        else
        {
            binding.cardSlotSelection.visibility = View.GONE
        }
    }

    private fun onReadyClass() {
        val preferences = PreferenceManager.getDefaultSharedPreferences(this)
        val isAutoConnectCamera = preferences.getBoolean(IPreferencePropertyAccessor.AUTO_CONNECT_TO_CAMERA, true)
        Log.v(TAG, "isAutoConnectCamera() : $isAutoConnectCamera")

        if (isAutoConnectCamera) {
            sceneUpdater?.changeCameraConnection()
        }
    }

    private fun vibrate() {
        val vibrator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val vibratorManager = getSystemService(VIBRATOR_MANAGER_SERVICE) as VibratorManager
            vibratorManager.defaultVibrator
        } else {
            @Suppress("DEPRECATION")
            getSystemService(VIBRATOR_SERVICE) as Vibrator
        }

        if (!vibrator.hasVibrator()) {
            Log.v(TAG, "No Vibrator available.")
            return
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            vibrator.vibrate(VibrationEffect.createOneShot(50, VibrationEffect.DEFAULT_AMPLITUDE))
        } else {
            @Suppress("DEPRECATION")
            vibrator.vibrate(50)
        }
    }

    override fun updateMessage(message: String, isBold: Boolean, isColor: Boolean, color: Int) {
        runOnUiThread {
            binding.message.text = message
            binding.message.typeface = if (isBold) Typeface.DEFAULT_BOLD else Typeface.DEFAULT
            binding.message.setTextColor(if (isColor) color else Color.DKGRAY)
        }
    }

    override fun setupSlotSelector(isEnable: Boolean, slotSelectionReceiver: ICardSlotSelectionReceiver?) {
        Log.v(TAG, "setupSlotSelector $isEnable")
        this.slotSelectionReceiver = slotSelectionReceiver
        runOnUiThread { setupCardSlotSelection(isEnable) }
    }

    override fun selectSlot(slotId: String) {
        Log.v(TAG, "selectSlot : $slotId")
    }

    override fun changedCardSlot(slotId: String) {
        Log.v(TAG, "changedCardSlot : $slotId")
        sceneUpdater?.reloadRemoteImageContents()
        vibrate()
    }

    companion object {
        private val TAG = MainActivity::class.java.simpleName
    }
}
